### PR TITLE
[types] support multiple signature schemes in transaction

### PR DESCRIPTION
**Pull request by sblackshear**
Implements end-end support for using multiple signature schemes in Libra transactions. There are currently two schemes: Ed25519 (as before) and MultiEd25519 (new) for K-of-N multisig (see 4d2a425283f1f23d4b30b85d5cefd6f8969ca8f6).
    
- Introduce `TransactionAuthenticator`, an abstraction of a signature scheme than knows how to convert its public key into an `AuthenticationKey` whose preimage can be stored in an on-chain account.
- The abstraction supports the two schemes mentioned above. New schemes can be added by simply extending the `enum` provided
- `AuthenticationKey` structure is now `sha3(public_key | scheme_id)` (whereas before it was just `sha3(public_key)`.
- Change code that relies on hardcoded addresses (which are derived from authentication keys)
- Generalize `SignedTransaction` to use an authenticator instead of an Ed25519 public key + signature, do associated refactoring.
    
Two notes:
- The client CLI has not yet been adapted to support multisig. That can come in a follow-up.
- The `AuthenticationKey` type should be migrated to transaction_authenticator.rs in the future
- I had to fix two small bugs in `MultiEd25519` to get my test to work.
  
## Motivation

Supporting multisig now and crypto agility in the long term.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

New end-to-end tests for multisig transactions, existing transaction tests all pass

### CODE REVIEW

File: types/src/transaction/authenticator.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            Self::MultiEd25519 { .. } => 1,
+        }
+    }
+
+    pub fn verify_signature(&self, message: &HashValue) -> Result<()> {
+        match self {
+            Self::Ed25519 {
+                public_key,
+                signature,
+            } => public_key.verify_signature(message, signature),
```
Comment (kchalkias) [inline types/src/transaction/authenticator.rs]: Both implement the `Signature` trait, so better use 
`signature.verify(message, public_key)` for consistency.

Comment (kchalkias) [reply types/src/transaction/authenticator.rs]: Also, as all of them are `VerifyingKeys`, do you need to distinguish between schemes? or just ensure the enum is correct and then always  `signature.verify`?

File: types/src/transaction/authenticator.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        }
+    }
+
+    /// Compute an account authentication using sha3(self.public_key | self.scheme_id)
+    pub fn authentication_key(&self) -> AuthenticationKey {
+        let mut bytes = match self {
+            Self::Ed25519 {
+                public_key,
+                signature,
+            } => public_key.to_bytes().to_vec(),
```
Comment (kchalkias) [inline types/src/transaction/authenticator.rs]: They are both `ValidKey`s so you can use the `to_bytes()` from there?

**Comment by sblackshear** [comment]
> Have we considered making SignedTransaction a generic SignedTransaction where T dictates the signature type? Also it is worth stating that given a Signature type T, one can obtain the "PublicKey/VerifyingKey" Type: https://github.com/libra/libra/blob/master/crypto/crypto/src/traits.rs#L193
> 
> Then in the enum for different type of Transactions (https://github.com/libra/libra/blob/master/types/src/transaction/mod.rs#L1071), you would have SignedTransaction SignedTransaction<Multi...>. All the magic would have to happen in an interface on SignedTransaction similar to what you're doing now.
> 
> More Explicitly
> 
> enum Transaction {
> UserTransaction(SignedTransaction),
> MultiSigUserTransaction(SignedTransaction),
> }
> 
> There might be a lot of weeds to pull out from that so you could always imagine using type assignment so that struct SignedTransaction is renamed to GenericSignedTransaction
> 
> pub type SignedTransaction = GenericSignedTransaction and maybe everything works without any additional modification.
> 
> If this sounds incoherent, it might be... but I can also walk through the idea in more details.

I went deep down the variant of this route we discussed offline, but it didn't end up making things simple for an annoying reason: Rust won't let you return `impl SomeTrait` if `SomeTrait` has associated types that refer to `Self`. Unfortunately, both `VerifyingKey` and `Signature` do this. So the `TransactionAuthenticator` has some repeated pattern matching, which is unfortunate but not the end of the world.

File: language/e2e-tests/src/tests/rotate_key.rs [hunk 3]
```diff
@@ -61,3 +67,65 @@ fn rotate_key() {
         );
     });
 }
+
+#[test]
```
Comment (kchalkias) [inline language/e2e-tests/src/tests/rotate_key.rs]: add `#[allow(clippy::redundant_clone)]`

**Pull request closed without merging**